include/snap/plugin.h

```diff
@@ -23,43 +23,120 @@ limitations under the License.
 
 namespace Plugin {
 
+/**
+ * Type is the plugin type
+ */
 enum Type {
-  collector,
-  processor,
-  publisher
+  Collector,
+  Processor,
+  Publisher
 };
 
-class PluginInterface {
+/**
+ * RpcType is the rpc protocol the plugin should use.
+ * Only GRPC is supported for C++.
+ */
+enum RpcType {
+  GRPC = 2,
+};
 
+/**
+ * Strategy is the routing and caching strategy this plugin requires.
+ * A routing and caching Strategy is the method which snapd uses to cache 
+ * metrics, and select the correct running instance of a plugin.
+ */
+enum Strategy {
+  /**
+   * LRU: Least Recently Used.
+   * The default strategy.
+   */
+  LRU,
+
+  /**
+   * Sticky: Sticky Running plugins have a 1:1 relationship to Tasks.
+   */
+  Sticky,
+
+  /** ConfigBased: Bases routing decisions on the incoming config.
+   * The config based strategy hashes the incoming config data and
+   * uses it as a key to select a running plugin.  It can be useful for plugins
+   * which maintain a connection to a database, for instance.
+   */
+  ConfigBased
+};
+
+/**
+ * Meta is the metadata about the plugin.
+ */
+struct Meta final {
   public:
-    virtual Config::Policy getConfigPolicy() = 0;
+    Meta(Type type, std::string name, int version) : type(type),
+                                                     name(name),
+                                                     version(version) {};
+
+    Type type;
+    std::string name;
+    int version;
+
+    // These members all have defaults.
+    RpcType rpc_type = RpcType::GRPC;
+    int concurrency_count = 5;
+    bool exclusive = false;
+    bool unsecure = false;
+    std::chrono::milliseconds cache_ttl = std::chrono::milliseconds(500);
+    Strategy strategy = Strategy::LRU;
 };
 
-class CollectorInterface : public PluginInterface {
+/**
+ * PluginInterface is the interface implemented by ALL plugins.
+ * Every plugin must implement get_config_policy.
+ */
+class PluginInterface {
+  public:
+    virtual ConfigPolicy get_config_policy() = 0;
+};
 
+/**
+ * The interface for a collector plugin.
+ * A Collector is the source.
+ * It is responsible for collecting metrics in the snap pipeline.
+ */
+class CollectorInterface : public PluginInterface {
   public:
```

Comment:
It works now because PluginInterface doesn't have a constructor. Maybe create placeholder ones, so you don't forget to initialize the parent as well?

Also, it's also good form to have a virtual destructor on the parent. A `delete` on a derived object (treated as PluginInterface) will leak data, if you don't have that in place.